### ISSUE TITLE
SetCurrentDirectory MAX_PATH

### ISSUE THREAD

**Issue by jborean93** [MicrosoftDocs/feedback#1441]
**🛑 IMPORTANT**: You can get your feedback addressed faster if you **use the comment section for the article in which you encountered a problem**.

**Link to article:**
https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setcurrentdirectory

**Problem:**
The docs for SetCurrentDirectory states

> In the ANSI version of this function, the name is limited to MAX_PATH characters. To extend this limit to 32,767 wide characters, call the Unicode version of the function and prepend "\?" to the path. For more information, see Naming a File.

This seems to indicate that it is possible to set a working directory to a path that exceeds `MAX_PATH` if you use the unicode version. This does not seem to be the case and no matter what I do I cannot change the current directory to a path that exceeds `MAX_PATH`.

PowerShell script to test this;

```
Add-Type -TypeDefinition @'
using System;
using System.Runtime.InteropServices;
using System.Text;

public class NativeMethods
{
    [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
    public static extern bool CreateDirectoryW(
        [MarshalAs(UnmanagedType.LPWStr)] string lpPathName,
        IntPtr lpSecurityAttributes);

    [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
    public static extern UInt32 GetCurrentDirectoryW(
        UInt32 nBufferLength,
        StringBuilder lpBuffer);

    [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
    public static extern bool SetCurrentDirectoryW(
        [MarshalAs(UnmanagedType.LPWStr)] string lpPathName);

    [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
    public static extern bool RemoveDirectoryW(
        [MarshalAs(UnmanagedType.LPWStr)] string lpPathName);
}
'@

Function Get-Win32ErrorMessage {
    [CmdletBinding()]
    Param (
        [Parameter(Mandatory=$true)][System.Int32]$ErrorCode
    )

    $exp = New-Object -TypeName System.ComponentModel.Win32Exception -ArgumentList $ErrorCode
    ("{0} (Win32 ErrorCode {1}, 0x{1:X8})" -f $exp.Message, $ErrorCode)
}

Function Get-CurrentDirectory {
    [OutputType([System.String])]
    [CmdletBinding()]

    $buffer = New-Object -TypeName System.Text.StringBuilder
    $length = [NativeMethods]::GetCurrentDirectoryW(0, $buffer)
    $buffer.EnsureCapacity($length) > $null

    $res = [NativeMethods]::GetCurrentDirectoryW(
        $length,
        $buffer
    ); $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()

    if ($res -eq 0) {
        throw "Failed to get current working directory: $(Get-Win32ErrorMessage -ErrorCode $err)"
    }

    return $buffer.ToString()
}

Function Remove-Directory {
    [CmdletBinding()]
    Param (
        [Parameter(Mandatory=$true)][System.String]$Path
    )

    $res = [NativeMethods]::RemoveDirectoryW($Path); $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
    if (-not $res) {
        throw "Failed to remove directory '$Path': $(Get-Win32ErrorMessage -ErrorCode $err)"
    }
}

Function Set-CurrentDirectory {
    [CmdletBinding()]
    Param (
        [Parameter(Mandatory=$true)][System.String]$Path
    )

    $res = [NativeMethods]::SetCurrentDirectoryW($Path); $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
    if (-not $res) {
        throw "Failed to set current directory to '$Path': $(Get-Win32ErrorMessage -ErrorCode $err)"
    }
}

Function Test-SetCurrentDirectory {
    [CmdletBinding()]
    Param (
        [Parameter(Mandatory=$true)][System.String]$Path
    )

    Write-Verbose -Message "Creating test directory at '$Path'"
    $res = [NativeMethods]::CreateDirectoryW(
        $Path,
        [System.IntPtr]::Zero
    ); $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
    if (-not $res) {
        throw "Failed to create directory '$Path': $(Get-Win32ErrorMessage -ErrorCode $err)"
    }

    try {
        Write-Verbose -Message "Getting current working directory"
        $orig_cwd = Get-CurrentDirectory
        Write-Verbose -Message "Original working directory: '$orig_cwd'"

        try {
            Write-Verbose -Message "Setting working directory to '$Path'"
            Set-CurrentDirectory -Path $Path

            Write-Verbose -Message "Getting new working directory after setting"
            $new_cwd = Get-CurrentDirectory
            Write-Verbose -Message "New working directory: '$new_cwd'"
        } finally {
            Set-CurrentDirectory -Path $orig_cwd
        }
    } finally {
        Remove-Directory -Path $Path
    }
}

$test_path = "C:\test folder"
if (Test-Path -LiteralPath $test_path) {
    Remove-Directory -Path $test_path
}
New-Item -Path $test_path -ItemType Directory > $null

try {
     # Test that a short path works
     Test-SetCurrentDirectory -Path "$test_path\short" -Verbose

     # Test that a short path with the NT prefix '\\?\' works
     Test-SetCurrentDirectory -Path "\\?\$test_path\short" -Verbose

     # Test a path that exceeds MAX_PATH, will fail on SetCurrentDirectoryW() with
     # The filename or extension is too long (Win32 ErrorCode 206, 0x000000CE)
     Test-SetCurrentDirectory -Path "\\?\$test_path\long - $("a" * 248)" -Verbose
} finally {
     Remove-Directory -Path $test_path
}
```

When running you get the following output

```
VERBOSE: Creating test directory at 'C:\test folder\short'
VERBOSE: Getting current working directory
VERBOSE: Original working directory: 'C:\Users\vagrant-domain'
VERBOSE: Setting working directory to 'C:\test folder\short'
VERBOSE: Getting new working directory after setting
VERBOSE: New working directory: 'C:\test folder\short'
VERBOSE: Creating test directory at '\\?\C:\test folder\short'
VERBOSE: Getting current working directory
VERBOSE: Original working directory: 'C:\Users\vagrant-domain'
VERBOSE: Setting working directory to '\\?\C:\test folder\short'
VERBOSE: Getting new working directory after setting
VERBOSE: New working directory: '\\?\C:\test folder\short'
VERBOSE: Creating test directory at '\\?\C:\test folder\long -
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaa'
VERBOSE: Getting current working directory
VERBOSE: Original working directory: 'C:\Users\vagrant-domain'
VERBOSE: Setting working directory to '\\?\C:\test folder\long -
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaa'
Failed to set current directory to '\\?\C:\test folder\long - aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa': The filename or extension is too long
(Win32 ErrorCode 206, 0x000000CE)
At line:9 char:9
+         throw "Failed to set current directory to '$Path': $(Get-Win3 ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Failed to set c...06, 0x000000CE):String) [], RuntimeException
    + FullyQualifiedErrorId : Failed to set current directory to '\\?\C:\test folder\long - aaaaaaaaaaaaaaaaaaaaaaaaaa
   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa': The
  filename or extension is too long (Win32 ErrorCode 206, 0x000000CE)
```

This shows that `SetCurrentDirectoryW()` fails when setting a path that exceeds `MAX_PATH` even with the `\\?\` prefix contrary to the docs. A quick Google search seems to indicate this is not possible so it would be good if the docs state that this is the case, or the proper method is documented (if it is possible).

I've tested the script on Server 2008 and 2019 and the versions in between.

**Comment by sudeepku** [comment]
@drewbatgit - could you take a look at this issue.

**Comment by drewbatgit** [comment]
There is another clause in the SetCurrentDirectory article that indicates the behavior has been updated in recent versions of Windows 10:

_Starting with Windows 10, version 1607, for the unicode version of this function (SetCurrentDirecotryW), you can opt-in to remove the MAX_PATH limitation. See the "Maximum Path Length Limitation" section of Naming Files, Paths, and Namespaces for details_.

The linked article has a section that details the changes and states that you can opt into the new behavior, which removes the MAX_PATH limitation by setting a registry key:

https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file#maximum-path-length-limitation

Do you want to take a look at that section of the article and see if that fixes your issue?

**Comment by jborean93** [comment]
@drewbatgit that works but unfortunately will only be useful for newer builds that support it. Perfectly fine if `SetCurrentDirectory` only works with long paths if that key is set but it would be good if the docs indicate that the `\\?\` method doesn't work.

**Comment by eryksun** [comment]
> @drewbatgit that works but unfortunately will only be useful for newer builds that support it. Perfectly fine if `SetCurrentDirectory` only works with long paths if that key is set but it would be good if the docs indicate that the `\\?\` method doesn't work.

The `SetCurrentDirectory` documentation didn't used to mention "\\\\?\\" paths  (check an Internet archive). The buffer for the current directory in the process parameters is and always has been statically limited to `MAX_PATH` characters, unless long paths are enabled in Windows 10. The system runtime library even has some nasty bugs if we set the current directory to a "\\\\?\\" path, so it's clear that no one ever intended for this to be supported. Some time a few years ago someone updated the docs to add this note that "\\\\?\\" paths can be used, but it's completely wrong.

Note that we need to be mindful when setting the current directory to a long path in Windows 10. `CreateProcess` does not support a current directory path length that exceeds `MAX_PATH - 2` characters (not counting the trailing backslash and null terminator). If the current directory path length is longer than this, `CreateProcess` will fail with an invalid-parameter error when we pass `lpCurrentDirectory` as `NULL` (i.e. inherit the current directory), though this is a normal and common value for this parameter.

**Comment by drewbatgit** [comment]
@eryksun You seem to have technical expertise in this area and I do not, so I'm inclined to just accept your feedback as-is. But I looked at the history of this API documentation and it looks like this text was added specifically at the request of the feature dev at the time. That doesn't mean it's correct, just that it was deliberate at the time and from a reliable source. It's possible that the statement is an incorrect phrasing of a legitimate feature, and therefore should be modified instead of removed. So I'm going to investigate this internally and will respond to the thread when I have more information.
